org.eclipse.sprotty/src/main/java/org/eclipse/sprotty/DefaultDiagramServer.java

```diff
@@ -360,13 +365,53 @@ private void doSubmitModel(SModelRoot newRoot, boolean update, Action cause) {
 					dispatch(new SetModelAction(newRoot));
 				}
 				lastSubmittedModelType = modelType;
+				updateSelection(newRoot, update, cause);
 				IModelUpdateListener listener = getModelUpdateListener();
 				if (listener != null) {
 					listener.modelSubmitted(newRoot, this);
 				}
 			}
 		}
 	}
+
+	private void updateSelection(SelectAction action) {
+		if (action.getDeselectedElementsIDs() != null) {
+			selectedElements.removeAll(action.getDeselectedElementsIDs());
+		}
+		if (action.getSelectedElementsIDs() != null) {
+			selectedElements.addAll(action.getSelectedElementsIDs());
+		}
+
+		fireSelectionChanged(action);
+	}
+
+	private void updateSelection(SelectAllAction action) {
+		if (action.isSelect()) {
+			selectedElements.clear();
+			selectedElements.addAll(new SModelIndex(getModel()).allIds());
+		} else {
+			selectedElements.clear();
+		}
+
+		fireSelectionChanged(action);
+	}
+
+	private void updateSelection(SModelRoot newRoot, boolean update, Action cause) {
+		if (update) {
+			selectedElements.retainAll(new SModelIndex(newRoot).allIds());
+		} else {
+			selectedElements.clear();
+		}
+
+		fireSelectionChanged(cause);
```

Comment:
Please avoid unnecessary selection change events, e.g.:

```java
int previousSize = selectedElements.size();
if (update) {
	selectedElements.retainAll(new SModelIndex(newRoot).allIds());
} else {
	selectedElements.clear();
}
if (selectedElements.size() != previousSize)
	fireSelectionChanged(cause);
```

Reply:
I considered this, but for the same reasons (breaking existing applications) I decided not to, because in the previous implementation an event was fired every time.

Reply:
I added some additional checks, but I'm unsure about firing event when the model changes. In a scenario where the model is replaced (not updated), we probably don't want to fire an event because we cleared the selection of the previous model?

Reply:
The previous implementation did not fire any selection change event for model updates, so I think it's good to check whether anything actually changed.